### PR TITLE
Added `cider-eval-sexp-to-point` for [Feature #2328]

### PR DESCRIPTION
**Pull request by zcaudate**
Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [X] The commits are consistent with our [contribution guidelines][1]
- [X] You've added tests (if possible) to cover your change(s)
- [X] All tests are passing (`make test`)
- [X] All code passes the linter (`make lint`) which is based on [`elisp-lint`](https://github.com/gonewest818/elisp-lint) and includes
  - [byte-compilation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte-Compilation.html), [`checkdoc`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips.html), [check-declare](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html), packaging metadata, indentation, and trailing whitespace checks.
- [X] You've updated the [changelog][3] (if adding/changing user-visible functionality)
- [X] You've updated the [user manual][4] (if adding/changing user-visible functionality)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][2] extremely useful.*

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md
[2]: https://cider.readthedocs.io/en/latest/hacking_on_cider/
[3]: https://github.com/clojure-emacs/cider/blob/master/CHANGELOG.md
[4]: https://github.com/clojure-emacs/cider/tree/master/doc

### CODE REVIEW

File: cider-interaction.el [hunk 1]
```diff
@@ -1440,6 +1440,22 @@ It constructs an expression to eval in the following manner:
...
+
+- It find the code between the point and the start of the sexp expression;
+- It balances this bit of code by closing the expression;
+- It evaluates the resulting code using `cider-interactive-eval'
+
+If invoked with OUTPUT-TO-CURRENT-BUFFER, print the result in the current buffer."
+  (interactive "P")
+  (let* ((beg-of-sexp (save-excursion (up-list) (backward-list) (point)))
+         (code (buffer-substring-no-properties beg-of-sexp (point)))
+         (code (concat code '(41))))
```
Comment (bbatsov) [inline cider-interaction.el]: Hmm, seems to me you hardcode a single closing paren here, which might be incorrect depending on the sexp, right?

Comment (bbatsov) [reply cider-interaction.el]: Here's a simple problem with the current solution - the current sexp might not be a list. Imagine you call this function in some vector or hash literal for instance. Why didn't you use the matching delimiter functionality which already exists?

**Comment by zcaudate** [comment]
I've added additional code to support sets as well.

**Comment by bbatsov** [comment]
The build is failing due to some problems with the docstring. See my inline comments.

File: cider-interaction.el [hunk 2]
```diff
@@ -1440,6 +1440,23 @@ It constructs an expression to eval in the following manner:
          (code (concat code (cider--calculate-closing-delimiters))))
     (cider-interactive-eval code)))
 
+(defun cider-eval-sexp-to-point (&optional  output-to-current-buffer)
```
Comment (bbatsov) [inline cider-interaction.el]: Why did you add this param in the first place? It makes no sense to me.

File: cider-interaction.el [hunk 3]
```diff
@@ -1440,6 +1440,23 @@ It constructs an expression to eval in the following manner:
...
+  "Evaluate the current sexp form up to point.
+It constructs an expression to eval in the following manner:
+
+- It find the code between the point and the start of the sexp expression;
+- It balances this bit of code by closing the expression;
+- It evaluates the resulting code using `cider-interactive-eval', working with C-u."
```
Comment (bbatsov) [inline cider-interaction.el]: Related to the first comment - no idea why you did this like that. You can simply write in `C-u` in a docstring, btw, and you have to refer explicitly to all params. I'd just remove the param.

File: cider-interaction.el [hunk 4]
```diff
@@ -1440,6 +1440,23 @@ It constructs an expression to eval in the following manner:
...
+  (interactive "P")
+  (let* ((beg-of-sexp (save-excursion (up-list) (backward-list) (point)))
+         (beg-delimiter (save-excursion (up-list) (backward-list) (char-after)))
+         (beg-set?  (save-excursion (up-list) (backward-list) (char-before)))
+         (code (buffer-substring-no-properties beg-of-sexp (point)))
+         (code (if (= beg-set? ?#) (concat (list beg-set?) code) code))
+         (code (concat code (list (cider--matching-delimiter beg-delimiter)))))
+    (cider-interactive-eval code
```
Comment (bbatsov) [inline cider-interaction.el]: This should be just a simple `cider-interactive-eval` to be consistent with other commands we have.

Comment (zcaudate) [reply cider-interaction.el]: outputting is really important (for me anyways) It's for inline editing:

out

File: cider-interaction.el (hunk 3 above)
Comment (zcaudate) [reply cider-interaction.el]: I didn't do the merge right. will resubmit

File: cider-interaction.el (hunk 4 above)
Comment (bbatsov) [reply cider-interaction.el]: I understand you're point but we need to be uniform in our command interface. There are dedicated commands for evaling something to comment, which is what you seem to want (.e.g. `cider-pprint-eval-last-sexp-to-comment` ). It seems to me we just need two command. The basic command that evals a sexp up to point and another command that wraps the first one and provides the pprint-to-comment functionality.

Comment (zcaudate) [reply cider-interaction.el]: well... I copied this function:

```
(defun cider-eval-last-sexp (&optional output-to-current-buffer)
  "Evaluate the expression preceding point.
If invoked with OUTPUT-TO-CURRENT-BUFFER, print the result in the current buffer."
  (interactive "P")
  (cider-interactive-eval nil
                          (when output-to-current-buffer (cider-eval-print-handler))
                          (cider-last-sexp 'bounds)))
```

the great thing is that C-x C-e outputs as a message whilst C-u C-x C-e outputs to the buffer.

Comment (bbatsov) [reply cider-interaction.el]: Oh, that's a great point. Turns out it's me who made a mistake by forgetting to include this in eval-defun-to-point. :-)

File: cider-mode.el [hunk 5]
```diff
@@ -233,6 +234,7 @@ Configure `cider-cljs-repl-types' to change the ClojureScript REPL to use for yo
     ["Insert ns form in REPL" cider-insert-ns-form-in-repl]
     ["Eval top-level sexp to comment" cider-eval-defun-to-comment]
     ["Eval top-level sexp and pretty-print to comment" cider-pprint-eval-defun-to-comment]
+    ["Eval top-level sexp to point" cider-eval-defun-to-point]
```
Comment (bbatsov) [inline cider-mode.el]: I already this command in the menu (a bit higher up).

File: cider-interaction.el [hunk 6]
```diff
@@ -1440,6 +1440,23 @@ It constructs an expression to eval in the following manner:
          (code (concat code (cider--calculate-closing-delimiters))))
     (cider-interactive-eval code)))
 
+(defun cider-eval-sexp-to-point (&optional  output-to-current-buffer)
+  "Evaluate the current sexp form up to point.
+It constructs an expression to eval in the following manner:
+
+- It finds the code between the point and the start of the sexp expression;
+- It balances this bit of code by closing the expression;
+- It evaluates the resulting code using `cider-interactive-eval', if invoked with OUTPUT-TO-CURRENT-BUFFER, insert as comment in the current buffer."
```
Comment (bbatsov) [inline cider-interaction.el]: I'd structure the docstring as in here - https://github.com/clojure-emacs/cider/commit/28bdc1f2897d351f8a1761e4cf121a1dd353a667

This avoid the very long line.
